packages/nextjs/src/utils/config/types.ts

```diff
@@ -0,0 +1,49 @@
+export { SentryCliPluginOptions as SentryWebpackPluginOptions } from '@sentry/webpack-plugin';
+
+export type ExportedNextConfig = NextConfigObject | NextConfigFunction;
+
+export type NextConfigFunction = (
+  phase: string,
+  defaults: { defaultConfig: { [key: string]: unknown } },
+) => NextConfigObject;
+
+export type NextConfigObject = {
+  // whether or not next should create source maps for browser code
+  // see: https://nextjs.org/docs/advanced-features/source-maps
+  productionBrowserSourceMaps?: boolean;
+  // custom webpack options
+  webpack?: WebpackConfigFunction;
+} & {
+  // other `next.config.js` options
+  [key: string]: unknown;
+};
+
+export type WebpackConfigObject = {
+  devtool?: string;
+  plugins?: Array<{ [key: string]: unknown }>;
+  entry: WebpackEntryProperty;
+  output: { filename: string; path: string };
+  target: string;
+  context: string;
+} & {
+  // other webpack options
+  [key: string]: unknown;
+};
+
+export type WebpackConfigFunction = (config: WebpackConfigObject, options: BuildContext) => WebpackConfigObject;
+
+// Information about the current build environment
+export type BuildContext = { dev: boolean; isServer: boolean; buildId: string };
+
+// For our purposes, the value for `entry` is either an object, or an asyc function which returns such an object
+export type WebpackEntryProperty = EntryPropertyObject | EntryPropertyFunction;
```

Comment:
nit: Can we move this below the type declarations for `EntryPropertyObject` and `EntryPropertyFunction`. Easier to read imo. Ditto with other types in this PR.

Reply:
Should really lint for this tbh

Reply:
Oh, actually, I did that on purpose. For me it's:

"What's the big thing we're talking about?" "This thing." "What's it made out of?" "These things."

vs

"Here's a thing." "Why is it important?" "Hang, I'll tell you in a sec. Here's another thing." "Okay, why is that one important?" "They're both important, because they make up this thing, which is really the point." "Wait, what were those things again?"

I did do a little reorganizing, though.